onnx/defs/gen_doc.py

```diff
@@ -32,126 +38,229 @@ def display_attr_type(v):
-              s += '</dl>\n'
-
-          # type constraints
-          s += '\n#### Type Constraints'
-          s += '\n\n'
-          if schema.type_constraints:
-              s += '<dl>\n'
-              for type_constraint in schema.type_constraints:
-                  allowedTypes = type_constraint.allowed_type_strs
-                  if (len(allowedTypes) > 0):
-                      allowedTypeStr = allowedTypes[0]
-                  for allowedType in allowedTypes[1:]:
-                      allowedTypeStr += ', ' + allowedType
-                  s += '<dt><tt>{}</tt> : {}</dt>\n'.format(type_constraint.type_param_str, allowedTypeStr)
-                  s += '<dd>{}</dd>\n'.format(type_constraint.description)
-              s += '</dl>\n'
-
-          s += '\n\n'
-
-          if op_type in SNIPPETS:
-              s += '#### Examples\n\n'
-              for summary, code in sorted(SNIPPETS[op_type]):
-                  s += '<details>\n'
-                  s += '<summary>{}</summary>\n\n'.format(summary)
-                  s += '```python\n{}\n```\n\n'.format(code)
-                  s += '</details>\n'
-                  s += '\n\n'
-          fout.write(s)
+        fout.write('## Operator Schemas\n')
+        fout.write(
+            "*This file is automatically generated from the\n"
+            "            [def files](/onnx/defs) via [this script](/onnx/defs/gen_doc.py).\n"
+            "            Do not modify directly and instead edit operator definitions.*\n")
+
+        # domain -> support level -> name -> [schema]
+        index = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+        for schema in defs.get_all_schemas_with_history():
+            index[schema.domain][int(schema.support_level)][schema.name].append(schema)
+
+        fout.write('\n')
+
+        # Table of contents
+        for domain, supportmap in sorted(index.items()):
+            if domain == 'ai.onnx.ml' and not ONNX_ML:
+                continue
+            elif ONNX_ML and domain != 'ai.onnx.ml':
+                continue
+
+            if domain:
+                s = '* {}\n'.format(domain)
+                domain_prefix = '{}.'.format(domain)
+            else:
+                s = '* ai.onnx (default)\n'
+                domain_prefix = ''
+            fout.write(s)
+
+            for _, namemap in sorted(supportmap.items()):
+                for n, unsorted_versions in sorted(namemap.items()):
+                    versions = sorted(unsorted_versions, key=lambda s: s.since_version)
+                    schema = versions[-1]
+                    s = '  * {}<a href="#{}">{}</a>\n'.format(
+                            support_level_str(schema.support_level),
+                            domain_prefix + n, domain_prefix + n)
+                    fout.write(s)
+
+        fout.write('\n')
+
+        for domain, supportmap in sorted(index.items()):
+            if domain == 'ai.onnx.ml' and not ONNX_ML:
+                continue
```

Comment:
elif ONNX_ML and domain != 'ai.onnx.ml':
                continue

should also be added, I think.

Reply:
Nice catch